### DISCUSSION TITLE
Using user-provided memory area as a memory pool with deallocation support?

### DISCUSSION THREAD

**Discussion by v1993** [foonathan/memory#163] [Q&A] [answered]
I've looked at documentation and examples, but was neither able to confirm nor deny if what I'm trying to do is possible.

I'd like to use a library to manage (array) allocation/deallocation within memory block provided by me, ideally with no exception-free error handling. The idea is to manage a memory-mapped region of GPU-backed buffer, allocating and deallocating memory for texture uploads on the fly. Is it possible with this library?

FWIW, GPU mapping is not really relevant to what library needs to support, since it behaves like any other memory area allocated externally to this library.

**Comment by foonathan** [comment, marked as answer]
It sounds like you want `memory_pool` with a custom `BlockOrRawAllocator`. The easiest way is to use a very simple `BlockAllocator` like the one given in that example: https://memory.foonathan.net/md_doc_concepts.html#concept_blockallocator.

**Reply by v1993** [reply]
Thank you for your answer! Is this how single-block allocator should be done? Docs aren't very clear on this.

```c++
class block_allocator
{
public:
    block_allocator(std::size_t block_size, void* base_address)
    : block_size_(block_size), base_address_(base_address) {}
    
    memory_block allocate_block()
    {
        if (in_use_)
        {
            return {nullptr, 0};
        }
        else
        {
            return {base_address_, block_size_};
        }
    }
    
    void deallocate_block(memory_block b)
    {
        in_use_ = false;
    }
    
    std::size_t next_block_size() const
    {
        if (in_use_)
        {
            return 0;
        }
        else
        {
            return block_size_;
        }
    }
    
private:
    std::size_t block_size_;
    void* base_address_;
    bool in_use_{false};
};
```

And then catch exceptions from `allocate_array` method of `memory_pool` to handle cases where allocation failed.

**Reply by foonathan** [reply]
Yeah, something like that. Although, I don't think the allocators can handle `next_block_size() == 0` (you'll get an assertion failure). I don't know whether that behavior is acceptable for you.

**Reply by v1993** [reply]
Should I just unconditionally return `block_size_` there, then?

I guess I can try to allocate memory-mapped blocks directly inside this class, but that'll complicate things significantly with trying to find out which buffer returned memory belongs to, so I'd rather not.

**Reply by foonathan** [reply]
No, lying about the next block size doesn't help either :D 

The question is: if allocate_block() is called twice, what should be done? The caller expects that to work, the allocator can't implement it. The only option is to terminate the program or throw an exception at that point.

**Reply by v1993** [reply]
Well, throwing is suboptimal, but it's a decidedly slow path (compared to loading directly into PBO) anyways.

In theory, I can just allocate normal memory (with malloc/free or new/delete or whatever) in that case and check in caller if returned pointer belongs to original memory-mapped range. However, will memory returned by the first call be preferred to extra allocated one?

**Reply by foonathan** [reply]
`memory_pool` maintains a list of free memory, which is initially everything in your first memory block. If the free list is empty, but more memory is requested, a second block is requested to fill the free list. When pool memory is deallocated, it is placed back onto the free list and memory is recycled FIFO or sorted in the case of an array memory pool.
